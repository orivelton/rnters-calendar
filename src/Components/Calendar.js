import React, { useState, useEffect } from 'react';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';
import { DateRangePicker } from 'react-dates';
import DateFormated from './DateFormated';
import apiGet from '../api/api';
import { formatDate, checkUnavailableDays } from '../helpers/formatDate';
import moment from 'moment';

const Calendar = () => {
  const [startDate, setStartDate] = useState(null);
  const [endDate, setEndDate] = useState(null);
  const [focusedInput, setFocusedInput] = useState(null);
  const [unavailable, setUnavailable] = useState(false);

  const [calendar, setCalendar] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const result = await apiGet();
      const {data: { attributes }} = result;
      setCalendar(attributes);
    };
     fetchData();
   }, []);

   const isDayBlocked = day => {
    const dayCalendar = formatDate(day);
    const { unavailable_periods: unavailableDays } = calendar;

    return checkUnavailableDays(unavailableDays, dayCalendar);
  }

  const hasBlockedInTheRange = (start, end) => {
    const dateFormat = 'DD/MM/YYYY';
    const diff = moment(end).diff(start, 'days') + 1;
    const { unavailable_periods: unavailableDays } = calendar;
    let hasBlocked = false

    for (let i = 0; i < diff; i++) {
      const checkDate = moment(start).add(i, 'd').format(dateFormat);
      const resultCheck = checkUnavailableDays(unavailableDays, checkDate);

      if(resultCheck) hasBlocked = true;
    }

    return hasBlocked;
  }

  const handleErrorRange = () => {
    setUnavailable(true);
    setStartDate(null);
    setEndDate(null);
  }
  
  return (
    <>
      <DateRangePicker
        noBorder
        regular
        withPortal
        startDate={startDate}
        endDate={endDate}
        startDateId="1"
        endDateId="2"
        onDatesChange={({ startDate, endDate }) => {
          setUnavailable(false);

          if(hasBlockedInTheRange(startDate, endDate)) {
            handleErrorRange();
          } else {
            setStartDate(startDate);
            setEndDate(endDate);
          }
        }}
        focusedInput={focusedInput}
        onFocusChange={focusedInput => setFocusedInput(focusedInput)}
        displayFormat="DD/MM/YYYY"
        numberOfMonths={1}
        isDayBlocked={(day) => isDayBlocked(day)}
      />
      
      {
        unavailable && <p className="alert-danger" role="alert">Selecione outra data! Temos agendamento para esse grupo de datas</p>
      }
      
      <DateFormated date={startDate} />
      <DateFormated date={endDate} />
    </>
  )
};

export default Calendar;