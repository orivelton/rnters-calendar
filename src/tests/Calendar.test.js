import React from 'react';
import renderer from 'react-test-renderer';
import Calendar from '../Components/Calendar';

it('Renders correctly', () => {
  const tree = renderer.create(<Calendar />).toJSON();
  expect(tree).toMatchSnapshot();
});
