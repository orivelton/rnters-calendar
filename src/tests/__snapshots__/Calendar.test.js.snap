// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
Array [
  <p>
    Escolha as datas
  </p>,
  <div
    className="DateRangePicker DateRangePicker_1"
  >
    <div
      className="DateRangePickerInput DateRangePickerInput_1"
    >
      <div
        className="DateInput DateInput_1"
      >
        <input
          aria-describedby="DateInput__screen-reader-message-1"
          aria-label="Início"
          autoComplete="off"
          className="DateInput_input DateInput_input_1 DateInput_input__regular DateInput_input__regular_2"
          disabled={false}
          id="1"
          name="1"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Início"
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
        <p
          className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
          id="DateInput__screen-reader-message-1"
        >
          Navigate forward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
        </p>
      </div>
      <div
        aria-hidden="true"
        className="DateRangePickerInput_arrow DateRangePickerInput_arrow_1"
        role="presentation"
      >
        <svg
          className="DateRangePickerInput_arrow_svg DateRangePickerInput_arrow_svg_1"
          focusable="false"
          viewBox="0 0 1000 1000"
        >
          <path
            d="M694 242l249 250c12 11 12 21 1 32L694 773c-5 5-10 7-16 7s-11-2-16-7c-11-11-11-21 0-32l210-210H68c-13 0-23-10-23-23s10-23 23-23h806L662 275c-21-22 11-54 32-33z"
          />
        </svg>
      </div>
      <div
        className="DateInput DateInput_1"
      >
        <input
          aria-describedby="DateInput__screen-reader-message-2"
          aria-label="Fim"
          autoComplete="off"
          className="DateInput_input DateInput_input_1 DateInput_input__regular DateInput_input__regular_2"
          disabled={false}
          id="2"
          name="2"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder="Fim"
          readOnly={false}
          required={false}
          type="text"
          value=""
        />
        <p
          className="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
          id="DateInput__screen-reader-message-2"
        >
          Navigate backward to interact with the calendar and select a date. Press the question mark key to get the keyboard shortcuts for changing dates.
        </p>
      </div>
    </div>
  </div>,
]
`;
